File: implementations/typescript/src/dict.ts
/**
 * Dictionary — 128 static + 127 session IDs for key compression.
 *
 * | Range       | Purpose                  |
 * |-------------|--------------------------|
 * | `0x00-0x7F` | Static dictionary (128)  |
 * | `0x80-0xFE` | Session dictionary (127) |
 * | `0xFF`      | Raw UTF-8 key (escape)   |
 *
 * Keys that appear in the static dictionary are encoded as 1 byte
 * instead of their full UTF-8 representation.
 *
 * Ported from Rust `src/dict.rs`.
 */

// ═══ Reserved IDs ══════════════════════════════════════════════════════════

/** Maximum static dictionary ID (exclusive). */
export const STATIC_MAX = 0x80;

/** Maximum session dictionary ID (exclusive). */
export const SESSION_MAX = 0xff;

/** Sentinel: key is not in the dictionary, sent as raw text. */
export const ID_RAW = 0xff;

/** Total number of usable dictionary entries (static + session). */
export const TOTAL_ENTRIES = 255;

// ═══ Static dictionary (128 entries, IDs 0x00..0x7F) ══════════════════════

const STATIC_DICT: (string | null)[] = new Array(STATIC_MAX).fill(null);

// Core MCP/RPC keys (0x00..0x0F)
STATIC_DICT[0x00] = "tool";
STATIC_DICT[0x01] = "arguments";
STATIC_DICT[0x02] = "result";
STATIC_DICT[0x03] = "error";
STATIC_DICT[0x04] = "id";
STATIC_DICT[0x05] = "name";
STATIC_DICT[0x06] = "description";
STATIC_DICT[0x07] = "content";
STATIC_DICT[0x08] = "text";
STATIC_DICT[0x09] = "type";
STATIC_DICT[0x0a] = "method";
STATIC_DICT[0x0b] = "params";
STATIC_DICT[0x0c] = "jsonrpc";
STATIC_DICT[0x0d] = "data";
STATIC_DICT[0x0e] = "code";
STATIC_DICT[0x0f] = "message";

// Input/output (0x10..0x1F)
STATIC_DICT[0x10] = "input";
STATIC_DICT[0x11] = "output";
STATIC_DICT[0x12] = "stream";
STATIC_DICT[0x13] = "uri";
STATIC_DICT[0x14] = "mimeType";
STATIC_DICT[0x15] = "encoding";
STATIC_DICT[0x16] = "language";
STATIC_DICT[0x17] = "title";
STATIC_DICT[0x18] = "value";
STATIC_DICT[0x19] = "key";
STATIC_DICT[0x1a] = "path";
STATIC_DICT[0x1b] = "version";
STATIC_DICT[0x1c] = "schema";
STATIC_DICT[0x1d] = "default";
STATIC_DICT[0x1e] = "required";
STATIC_DICT[0x1f] = "properties";

// Resources & tools (0x20..0x2F)
STATIC_DICT[0x20] = "resources";
STATIC_DICT[0x21] = "tools";
STATIC_DICT[0x22] = "prompts";
STATIC_DICT[0x23] = "resource";
STATIC_DICT[0x24] = "prompt";
STATIC_DICT[0x25] = "handler";
STATIC_DICT[0x26] = "capabilities";
STATIC_DICT[0x27] = "permissions";
STATIC_DICT[0x28] = "scope";
STATIC_DICT[0x29] = "tags";
STATIC_DICT[0x2a] = "category";
STATIC_DICT[0x2b] = "icon";
STATIC_DICT[0x2c] = "metadata";
STATIC_DICT[0x2d] = "timestamp";
STATIC_DICT[0x2e] = "status";
STATIC_DICT[0x2f] = "progress";

// Errors & status (0x30..0x3F)
STATIC_DICT[0x30] = "severity";
STATIC_DICT[0x31] = "details";
STATIC_DICT[0x32] = "cause";
STATIC_DICT[0x33] = "stack";
STATIC_DICT[0x34] = "line";
STATIC_DICT[0x35] = "column";
STATIC_DICT[0x36] = "source";
STATIC_DICT[0x37] = "retry";
STATIC_DICT[0x38] = "timeout";
STATIC_DICT[0x39] = "limit";
STATIC_DICT[0x3a] = "offset";
STATIC_DICT[0x3b] = "count";
STATIC_DICT[0x3c] = "total";
STATIC_DICT[0x3d] = "page";
STATIC_DICT[0x3e] = "cursor";
STATIC_DICT[0x3f] = "next";

// LLM / AI (0x40..0x4F)
STATIC_DICT[0x40] = "model";
STATIC_DICT[0x41] = "provider";
STATIC_DICT[0x42] = "temperature";
STATIC_DICT[0x43] = "max_tokens";
STATIC_DICT[0x44] = "stop";
STATIC_DICT[0x45] = "frequency_penalty";
STATIC_DICT[0x46] = "presence_penalty";
STATIC_DICT[0x47] = "top_p";
STATIC_DICT[0x48] = "logprobs";
STATIC_DICT[0x49] = "user";
STATIC_DICT[0x4a] = "system";
STATIC_DICT[0x4b] = "assistant";
STATIC_DICT[0x4c] = "function";
STATIC_DICT[0x4d] = "tool_calls";
STATIC_DICT[0x4e] = "finish_reason";
STATIC_DICT[0x4f] = "usage";

// HTTP / Web (0x50..0x5F)
STATIC_DICT[0x50] = "url";
STATIC_DICT[0x51] = "http_method";
STATIC_DICT[0x52] = "headers";
STATIC_DICT[0x53] = "body";
STATIC_DICT[0x54] = "query";
STATIC_DICT[0x55] = "http_status";
STATIC_DICT[0x56] = "cookie";
STATIC_DICT[0x57] = "session";
STATIC_DICT[0x58] = "token";
STATIC_DICT[0x59] = "auth";
STATIC_DICT[0x5a] = "redirect";
STATIC_DICT[0x5b] = "host";
STATIC_DICT[0x5c] = "port";
STATIC_DICT[0x5d] = "origin";
STATIC_DICT[0x5e] = "referrer";
STATIC_DICT[0x5f] = "agent";

// File System (0x60..0x6F)
STATIC_DICT[0x60] = "filename";
STATIC_DICT[0x61] = "directory";
STATIC_DICT[0x62] = "extension";
STATIC_DICT[0x63] = "size";
STATIC_DICT[0x64] = "modified";
STATIC_DICT[0x65] = "created";
STATIC_DICT[0x66] = "accessed";
STATIC_DICT[0x67] = "mode";
STATIC_DICT[0x68] = "owner";
STATIC_DICT[0x69] = "group";
STATIC_DICT[0x6a] = "symlink";
STATIC_DICT[0x6b] = "binary";
STATIC_DICT[0x6c] = "base64";
STATIC_DICT[0x6d] = "hash";
STATIC_DICT[0x6e] = "algorithm";
STATIC_DICT[0x6f] = "chunk";

// Operations (0x70..0x7F)
STATIC_DICT[0x70] = "execute";
STATIC_DICT[0x71] = "read";
STATIC_DICT[0x72] = "write";
STATIC_DICT[0x73] = "delete";
STATIC_DICT[0x74] = "update";
STATIC_DICT[0x75] = "create";
STATIC_DICT[0x76] = "search";
STATIC_DICT[0x77] = "list";
STATIC_DICT[0x78] = "get";
STATIC_DICT[0x79] = "set";
STATIC_DICT[0x7a] = "watch";
STATIC_DICT[0x7b] = "subscribe";
STATIC_DICT[0x7c] = "notify";
STATIC_DICT[0x7d] = "cancel";
STATIC_DICT[0x7e] = "pause";
STATIC_DICT[0x7f] = "resume";

// ═══ Reverse lookup (lazily-built O(1) HashMap) ═══════════════════════════

let reverseMap: Map<string, number> | null = null;

function getReverseMap(): Map<string, number> {
  if (!reverseMap) {
    reverseMap = new Map();
    for (let i = 0; i < STATIC_MAX; i++) {
      const key = STATIC_DICT[i];
      if (key !== null) {
        reverseMap.set(key, i);
      }
    }
  }
  return reverseMap;
}

// ═══ Public API ════════════════════════════════════════════════════════════

/**
 * Resolve a dictionary ID to its key string.
 * Returns `null` if the ID is not in the static dictionary.
 * ID 0xFF (`ID_RAW`) always returns `null` (meaning "inline text").
 */
export function resolveDictId(id: number): string | null {
  if (id < STATIC_MAX) {
    return STATIC_DICT[id];
  }
  // Session dictionary not yet implemented
  return null;
}

/**
 * O(1) reverse lookup: find the static dictionary ID for a key string.
 * Uses a lazily-built `Map` for constant-time lookup.
 *
 * Always prefer this over linear scan in hot paths (serialization, compression).
 */
export function lookupDictId(key: string): number | null {
  return getReverseMap().get(key) ?? null;
}

File: implementations/csharp/Dict.cs
/// <summary>
/// LUMEN static dictionary — 128 predefined key mappings (0x00–0x7F).
/// Mirrors implementations/rust/src/dict.rs exactly.
/// </summary>
namespace Lumen;

public static class Dict
{
    /// <summary>Raw UTF-8 key sentinel.</summary>
    public const byte ID_RAW = 0xFF;

    /// <summary>Maximum static dictionary ID + 1.</summary>
    public const int STATIC_MAX = 128;

    /// <summary>Maximum session dictionary IDs.</summary>
    public const int SESSION_MAX = 127;

    /// <summary>Total entries (static + session).</summary>
    public const int TOTAL_ENTRIES = 255;

    private static readonly string[] _byId = new string[STATIC_MAX];
    private static readonly Dictionary<string, byte> _byKey = new(STATIC_MAX);

    // Session dictionary (0x80..0xFE, 127 dynamic slots)
    private static readonly string?[] _sessionForward = new string?[SESSION_MAX];
    private static readonly Dictionary<string, byte> _sessionReverse = new(SESSION_MAX);

    static Dict()
    {
        var entries = new (byte id, string key)[]
        {
            (0x00, "tool"), (0x01, "arguments"), (0x02, "result"), (0x03, "error"),
            (0x04, "id"), (0x05, "name"), (0x06, "description"), (0x07, "content"),
            (0x08, "text"), (0x09, "type"), (0x0A, "method"), (0x0B, "params"),
            (0x0C, "jsonrpc"), (0x0D, "data"), (0x0E, "code"), (0x0F, "message"),
            (0x10, "input"), (0x11, "output"), (0x12, "stream"), (0x13, "uri"),
            (0x14, "mimeType"), (0x15, "encoding"), (0x16, "language"), (0x17, "title"),
            (0x18, "value"), (0x19, "key"), (0x1A, "path"), (0x1B, "version"),
            (0x1C, "schema"), (0x1D, "default"), (0x1E, "required"), (0x1F, "properties"),
            (0x20, "resources"), (0x21, "tools"), (0x22, "prompts"), (0x23, "resource"),
            (0x24, "prompt"), (0x25, "handler"), (0x26, "capabilities"), (0x27, "permissions"),
            (0x28, "scope"), (0x29, "tags"), (0x2A, "category"), (0x2B, "icon"),
            (0x2C, "metadata"), (0x2D, "timestamp"), (0x2E, "status"), (0x2F, "progress"),
            (0x30, "severity"), (0x31, "details"), (0x32, "cause"), (0x33, "stack"),
            (0x34, "line"), (0x35, "column"), (0x36, "source"), (0x37, "retry"),
            (0x38, "timeout"), (0x39, "limit"), (0x3A, "offset"), (0x3B, "count"),
            (0x3C, "total"), (0x3D, "page"), (0x3E, "cursor"), (0x3F, "next"),
            (0x40, "model"), (0x41, "provider"), (0x42, "temperature"), (0x43, "max_tokens"),
            (0x44, "stop"), (0x45, "frequency_penalty"), (0x46, "presence_penalty"), (0x47, "top_p"),
            (0x48, "logprobs"), (0x49, "user"), (0x4A, "system"), (0x4B, "assistant"),
            (0x4C, "function"), (0x4D, "tool_calls"), (0x4E, "finish_reason"), (0x4F, "usage"),
            (0x50, "url"), (0x51, "http_method"), (0x52, "headers"), (0x53, "body"),
            (0x54, "query"), (0x55, "http_status"), (0x56, "cookie"), (0x57, "session"),
            (0x58, "token"), (0x59, "auth"), (0x5A, "redirect"), (0x5B, "host"),
            (0x5C, "port"), (0x5D, "origin"), (0x5E, "referrer"), (0x5F, "agent"),
            (0x60, "filename"), (0x61, "directory"), (0x62, "extension"), (0x63, "size"),
            (0x64, "modified"), (0x65, "created"), (0x66, "accessed"), (0x67, "mode"),
            (0x68, "owner"), (0x69, "group"), (0x6A, "symlink"), (0x6B, "binary"),
            (0x6C, "base64"), (0x6D, "hash"), (0x6E, "algorithm"), (0x6F, "chunk"),
            (0x70, "execute"), (0x71, "read"), (0x72, "write"), (0x73, "delete"),
            (0x74, "update"), (0x75, "create"), (0x76, "search"), (0x77, "list"),
            (0x78, "get"), (0x79, "set"), (0x7A, "watch"), (0x7B, "subscribe"),
            (0x7C, "notify"), (0x7D, "cancel"), (0x7E, "pause"), (0x7F, "resume"),
        };

        foreach (var (id, key) in entries)
        {
            _byId[id] = key;
            _byKey[key] = id;
        }
    }

    /// <summary>Look up a key → dictionary ID. Checks static then session dict. Returns null if not found.</summary>
    public static byte? LookupId(string key)
    {
        if (_byKey.TryGetValue(key, out var id))
            return id;
        return _sessionReverse.TryGetValue(key, out var sid) ? sid : null;
    }

    /// <summary>Resolve a dictionary ID → key string. Checks static then session. Returns null if out of range.</summary>
    public static string? ResolveId(byte id)
    {
        if (id < STATIC_MAX)
            return _byId[id];
        if (id < 0xFF)
            return _sessionForward[id - STATIC_MAX];
        return null;
    }

    // ═══ Session dictionary (0x80..0xFE, 127 dynamic slots) ═════════════════

    /// <summary>Register a key in the session dictionary at a specific ID. Returns true on success.</summary>
    public static bool RegisterSessionKey(string key, byte id)
    {
        if (id < STATIC_MAX || id >= 0xFF)
            return false;
        int idx = id - STATIC_MAX;
        var old = _sessionForward[idx];
        if (old is not null)
            _sessionReverse.Remove(old);
        _sessionForward[idx] = key;
        _sessionReverse[key] = id;
        return true;
    }

    /// <summary>Remove a key from the session dictionary by ID.</summary>
    public static void UnregisterSessionKey(byte id)
    {
        if (id < STATIC_MAX || id >= 0xFF)
            return;
        int idx = id - STATIC_MAX;
        var key = _sessionForward[idx];
        if (key is not null)
        {
            _sessionReverse.Remove(key);
            _sessionForward[idx] = null;
        }
    }

    /// <summary>Initialize the session dictionary from (id, key) pairs.</summary>
    public static void InitSessionDict(IEnumerable<(byte id, string key)> entries)
    {
        ClearSessionDict();
        foreach (var (id, key) in entries)
            RegisterSessionKey(key, id);
    }

    /// <summary>Remove all session dictionary entries.</summary>
    public static void ClearSessionDict()
    {
        Array.Clear(_sessionForward, 0, _sessionForward.Length);
        _sessionReverse.Clear();
    }

    /// <summary>Number of registered session entries.</summary>
    public static int SessionDictSize => _sessionReverse.Count;
}
